[Bug] Imports of > 1000 notes fail

```
java.lang.RuntimeException: An error occurred while executing doInBackground()
at android.os.AsyncTask$4.done(AsyncTask.java:399)
at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
at java.util.concurrent.FutureTask.run(FutureTask.java:271)
at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:289)
at java.util.concurrent.ThreadPoolExecutor.processTask(ThreadPoolExecutor.java:1187)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
at java.lang.Thread.run(Thread.java:929)
Caused by: net.ankiweb.rsdroid.BackendFatalError: panic with no info
at net.ankiweb.rsdroid.RustBackendImpl.validateResult(RustBackendImpl.java:2)
at net.ankiweb.rsdroid.BackendV1Impl.getPage(BackendV1Impl.java:6)
at net.ankiweb.rsdroid.BackendMutex.getPage(BackendMutex.java:2)
at net.ankiweb.rsdroid.database.Session.getPage(Session.java:1)
at net.ankiweb.rsdroid.database.StreamingProtobufSQLiteCursor.getNextPage(StreamingProtobufSQLiteCursor.java:3)
at net.ankiweb.rsdroid.database.StreamingProtobufSQLiteCursor.moveToNext(StreamingProtobufSQLiteCursor.java:2)
at com.ichi2.libanki.Collection._fieldData(Collection.java:3)
at com.ichi2.libanki.Collection.updateFieldCache(Collection.java:5)
at com.ichi2.libanki.Collection.updateFieldCache(Collection.java:2)
at com.ichi2.libanki.importer.Anki2Importer._importNotes(Anki2Importer.java:69)
at com.ichi2.libanki.importer.Anki2Importer._import(Anki2Importer.java:11)
at com.ichi2.libanki.importer.Anki2Importer.run(Anki2Importer.java:3)
at com.ichi2.libanki.importer.AnkiPackageImporter.run(AnkiPackageImporter.java:58)
at com.ichi2.async.CollectionTask$ImportAdd.task(CollectionTask.java:6)
at com.ichi2.async.CollectionTask$ImportAdd.task(CollectionTask.java:1)
at com.ichi2.async.CollectionTask.actualDoInBackground(CollectionTask.java:14)
at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:2)
at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:1)
at android.os.AsyncTask$3.call(AsyncTask.java:378)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
... 5 more
```

https://github.com/david-allison-1/Anki-Android-Backend/blob/edbac00d3c3992796400b68f81ef4bef8c1a2a15/rslib-bridge/src/lib.rs#L448-L458

Logs:

```
IntentHandler/ User requested to view a file
Intent: Intent { dat=[REDACTED] flg=0x43 }. Data: none
Attempting to read content from intent.
AnkiActivity::onStart - DeckPicker
AnkiActivity::onResume - DeckPicker
Dispatching persistent message: 2
Handling Message: ImportAddDialog
Updating deck list UI
uncompress collection.anki2
uncompress media
Opening rust backend. Server: false. Langs: 'en', path:
Opening Collection: '/storage/emulated/0/AnkiDroid/tmpzip/collection.anki2' '' '' ''
Opening Collection Log
Rust: transaction begin
flush - Saving information to DB...
Rust: executeGetRowsAffected UPDATE col SET ls=?,crt=?,dty=?,mod=?,scm=?,usn=?,conf=?
Rust: transaction commit
ExecuteCommand: closeCollection
Closing Collection Log
Collection closed
Opening rust backend. Server: false. Langs: 'en', path:
Opening Collection: '/storage/emulated/0/AnkiDroid/tmpzip/collection.anki2' '' '' ''
Opening Collection Log
Rust: transaction begin
Preparing Import
Importing notes
Rust: transaction rollback
DB/ Not in a transaction. Cannot end transaction.
ExecuteCommand: closeCollection
Closing Collection Log
Collection closed
```

```
2021-01-26 14:42:16.860 31526-31924/com.ichi2.anki E/com.ichi2.anki: [qarth_debug:]  get PatchStore::createDisableExceptionQarthFile method fail.
2021-01-26 14:42:16.869 31526-31924/com.ichi2.anki E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #2
    Process: com.ichi2.anki, PID: 31526
    java.lang.RuntimeException: An error occurred while executing doInBackground()
        at android.os.AsyncTask$3.done(AsyncTask.java:355)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:246)
        at java.util.concurrent.ThreadPoolExecutor.processTask(ThreadPoolExecutor.java:1187)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:784)
     Caused by: net.ankiweb.rsdroid.BackendFatalError: panic with no info
        at net.ankiweb.rsdroid.RustBackendImpl.validateResult(RustBackendImpl.java:35)
        at net.ankiweb.rsdroid.BackendV1Impl.getPage(BackendV1Impl.java:284)
        at net.ankiweb.rsdroid.BackendMutex.getPage(BackendMutex.java:145)
        at net.ankiweb.rsdroid.database.Session.getPage(Session.java:90)
        at net.ankiweb.rsdroid.database.StreamingProtobufSQLiteCursor.getNextPage(StreamingProtobufSQLiteCursor.java:66)
        at net.ankiweb.rsdroid.database.StreamingProtobufSQLiteCursor.moveToNext(StreamingProtobufSQLiteCursor.java:97)
        at com.ichi2.libanki.importer.Anki2Importer._importCards(Anki2Importer.java:557)
        at com.ichi2.libanki.importer.Anki2Importer._import(Anki2Importer.java:159)
        at com.ichi2.libanki.importer.Anki2Importer.run(Anki2Importer.java:115)
        at com.ichi2.libanki.importer.AnkiPackageImporter.run(AnkiPackageImporter.java:148)
        at com.ichi2.async.CollectionTask$ImportAdd.task(CollectionTask.java:1314)
        at com.ichi2.async.CollectionTask$ImportAdd.task(CollectionTask.java:1299)
        at com.ichi2.async.CollectionTask.actualDoInBackground(CollectionTask.java:222)
        at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:185)
        at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:108)
        at android.os.AsyncTask$2.call(AsyncTask.java:334)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:246) 
        at java.util.concurrent.ThreadPoolExecutor.processTask(ThreadPoolExecutor.java:1187) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:784) 
2021-01-26 14:42:16.870 31526-31924/com.ichi2.anki D/UsageAnalytics: sendAnalyticsException() description/fatal: net.ankiweb.rsdroid.BackendFatalError: panic with no info/true
2021-01-26 14:42:16.956 31526-31924/com.ichi2.anki I/ACRA: ACRA is disabled for com.ichi2.anki - forwarding uncaught Exception on to default ExceptionHandler
2021-01-26 14:42:16.980 31526-31924/com.ichi2.anki I/Process: Sending signal. PID: 31526 SIG: 9
```

Second cause: importing uses nested SQL cursors, which are not supported yet: https://github.com/david-allison-1/Anki-Android-Backend/issues/23

--- a/AnkiDroid/build.gradle
+++ b/AnkiDroid/build.gradle
@@ -242,7 +242,7 @@ dependencies {
     implementation 'androidx.webkit:webkit:1.4.0'
 
     // == Rust conversion (from Anki-Android-Backend on GitHub) ==
-    String backendVersion = "0.1.0" // We want both testing and implementation on the same version
+    String backendVersion = "0.1.1" // We want both testing and implementation on the same version
     // build with ./gradlew rsdroid:assembleRelease
     // In my experience, using `files()` currently requires a reindex operation, which is slow.
     // implementation files("C:\\GitHub\\Rust-Test\\rsdroid\\build\\outputs\\aar\\rsdroid-release.aar")

--- a/AnkiDroid/build.gradle
+++ b/AnkiDroid/build.gradle
@@ -246,6 +246,7 @@ dependencies {
     // build with ./gradlew rsdroid:assembleRelease
     // In my experience, using `files()` currently requires a reindex operation, which is slow.
     // implementation files("C:\\GitHub\\Rust-Test\\rsdroid\\build\\outputs\\aar\\rsdroid-release.aar")
+    implementation 'com.google.protobuf:protobuf-java:3.13.0' // This is required when loading from a file
     implementation "io.github.david-allison-1:anki-android-backend:$backendVersion"
     // build with ./gradlew rsdroid-testing:assembleRelease
     // RobolectricTest.java: replace RustBackendLoader.init(); with RustBackendLoader.loadRsdroid(path);

--- a/AnkiDroid/build.gradle
+++ b/AnkiDroid/build.gradle
@@ -242,7 +242,7 @@ dependencies {
     implementation 'androidx.webkit:webkit:1.4.0'
 
     // == Rust conversion (from Anki-Android-Backend on GitHub) ==
-    String backendVersion = "0.1.1" // We want both testing and implementation on the same version
+    String backendVersion = "0.1.2" // We want both testing and implementation on the same version
     // build with ./gradlew rsdroid:assembleRelease
     // In my experience, using `files()` currently requires a reindex operation, which is slow.
     // implementation files("C:\\GitHub\\Rust-Test\\rsdroid\\build\\outputs\\aar\\rsdroid-release.aar")

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java
@@ -2170,6 +2170,10 @@ public class Collection {
         return mDroidBackend.isUsingRustBackend();
     }
 
+    public DroidBackend getBackend() {
+        return mDroidBackend;
+    }
+
     /** Allows a mock db to be inserted for testing */
     @VisibleForTesting
     public void setDb(DB database) {

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/backend/DroidBackend.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/backend/DroidBackend.java
@@ -18,6 +18,8 @@ package com.ichi2.libanki.backend;
 
 import com.ichi2.libanki.DB;
 
+import androidx.annotation.VisibleForTesting;
+
 /**
  * Interface to the rust backend listing all currently supported functionality.
  */
@@ -29,4 +31,7 @@ public interface DroidBackend {
     boolean databaseCreationCreatesSchema();
 
     boolean isUsingRustBackend();
+
+    @VisibleForTesting(otherwise = VisibleForTesting.NONE)
+    void debugEnsureNoOpenPointers();
 }

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/backend/JavaDroidBackend.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/backend/JavaDroidBackend.java
@@ -50,4 +50,10 @@ public class JavaDroidBackend implements DroidBackend {
     public boolean isUsingRustBackend() {
         return false;
     }
+
+
+    @Override
+    public void debugEnsureNoOpenPointers() {
+        // no-op
+    }
 }

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/backend/RustDroidBackend.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/backend/RustDroidBackend.java
@@ -21,10 +21,13 @@ import com.ichi2.libanki.DB;
 import net.ankiweb.rsdroid.BackendFactory;
 import net.ankiweb.rsdroid.BackendUtils;
 
+import BackendProto.AdBackend;
 import timber.log.Timber;
 
 /** The Backend in Rust */
 public class RustDroidBackend implements DroidBackend {
+    public static final int UNUSED_VALUE = 0;
+
     // I think we can change this to BackendV1 once new DB() accepts it.
     private final BackendFactory mBackend;
 
@@ -55,5 +58,15 @@ public class RustDroidBackend implements DroidBackend {
     @Override
     public boolean isUsingRustBackend() {
         return true;
+    }
+
+
+    @Override
+    public void debugEnsureNoOpenPointers() {
+        AdBackend.DebugActiveDatabaseSequenceNumbersOut result = mBackend.getBackend().debugActiveDatabaseSequenceNumbers(UNUSED_VALUE);
+        if (result.getSequenceNumbersCount() > 0) {
+            String numbers = result.getSequenceNumbersList().toString();
+            throw new IllegalStateException("Contained unclosed sequence numbers: " + numbers);
+        }
     }
 }

--- a/AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.java
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.java
@@ -145,6 +145,9 @@ public class RobolectricTest {
         controllersForCleanup.clear();
 
         try {
+            if (CollectionHelper.getInstance().colIsOpen()) {
+                CollectionHelper.getInstance().getCol(getTargetContext()).getBackend().debugEnsureNoOpenPointers();
+            }
             // If you don't tear down the database you'll get unexpected IllegalStateExceptions related to connections
             CollectionHelper.getInstance().closeCollection(false, "RoboelectricTest: End");
         } catch (BackendException ex) {

